test/arithmetic/multiplier_test.dart

```diff
@@ -0,0 +1,221 @@
+    final multiplyOnly = mod is MultiplyOnly;
+    await mod.build();
+    final signed = mod.signed;
+    for (var i = 0; i < iterations; i++) {
+      final bA = signed
+          ? Random().nextLogicValue(width: width).toBigInt().toSigned(width)
+          : Random().nextLogicValue(width: width).toBigInt().toUnsigned(width);
+      final bB = signed
+          ? Random().nextLogicValue(width: width).toBigInt().toSigned(width)
+          : Random().nextLogicValue(width: width).toBigInt().toUnsigned(width);
+      final bC = multiplyOnly
+          ? BigInt.zero
+          : signed
+              ? Random().nextLogicValue(width: width).toBigInt().toSigned(width)
+              : Random()
+                  .nextLogicValue(width: width)
+                  .toBigInt()
+                  .toUnsigned(width);
+      checkMultiplyAccumulate(mod, bA, bB, bC);
+    }
+  });
+}
+
+// Exhaustive testing of a mutiplier or multiplier/accumulate unit
+void testMultiplyAccumulateExhaustive(
+    int width, MultiplyAccumulate Function(Logic a, Logic b, Logic c) fn) {
+  final a = Logic(name: 'a', width: width);
+  final b = Logic(name: 'b', width: width);
+  final c = Logic(name: 'c', width: 2 * width);
+  final mod = fn(a, b, c);
+  test('exhaustive_${mod.name}_S${mod.signed}_W$width', () async {
+    await mod.build();
+    final signed = mod.signed;
+    final multiplyOnly = mod is MultiplyOnly;
+
+    final cLimit = multiplyOnly ? 1 : (1 << (2 * width));
+
+    for (var aa = 0; aa < (1 << width); ++aa) {
+      for (var bb = 0; bb < (1 << width); ++bb) {
+        for (var cc = 0; cc < cLimit; ++cc) {
+          final bA = signed
+              ? BigInt.from(aa).toSigned(width)
+              : BigInt.from(aa).toUnsigned(width);
+          final bB = signed
+              ? BigInt.from(bb).toSigned(width)
+              : BigInt.from(bb).toUnsigned(width);
+          final bC = multiplyOnly
+              ? BigInt.zero
+              : signed
+                  ? BigInt.from(cc).toSigned(2 * width)
+                  : BigInt.from(cc).toUnsigned(2 * width);
+
+          checkMultiplyAccumulate(mod, bA, bB, bC);
+        }
+      }
+    }
+  });
+}
+
+void main() {
+  tearDown(() async {
+    await Simulator.reset();
+  });
+
+  // Use MAC tester for Multiply
+
+  // First curry the Multiplier
+  Multiplier currySignedMultiplier(Logic a, Logic b) =>
+      CompressionTreeMultiplier(a, b, 4, KoggeStone.new, signed: true);
+
+  Multiplier curryUnsignedMultiplier(Logic a, Logic b) =>
+      CompressionTreeMultiplier(a, b, 4, KoggeStone.new, signed: true);
+
+  // Now treat the multiplier as a MAC with a zero input addend [c]
+  MultiplyAccumulate currySignedMultiplierAsMAC(Logic a, Logic b, Logic c) =>
+      MultiplyOnly(a, b, c, currySignedMultiplier);
+
+  MultiplyAccumulate curryUnsignedMultiplierAsMAC(Logic a, Logic b, Logic c) =>
+      MultiplyOnly(a, b, c, curryUnsignedMultiplier);
+
```

Comment:
this is some really beautiful testing automation!

Reply:
Again, another area where I feel like I am getting the hang of Dart and can do this kind of thing much faster than I was able to in Chisel.  The only place were I felt Scala was a little cleaner was passing back values from control structures, other than that I have been more than pleasantly surprised how easily I could mix in functional style where it helps.